doc/design/optimizer.md

```diff
@@ -0,0 +1,88 @@
+## Optimizer Design
+Optimizer is used to update parameter according to parameter gradient. 
+
+#### A typical training process is:
+
+1. run forward to calculate activation using data and parameter.
+1. run backward to calculate the gradient of activation and parameter using cost, activation, and parameter.
+1. apply the gradient to the corresponding parameter.
+
+User write code to describe the forward process, Optimizer should help implicitly add the process of backward and update according to forward process and other configuration.
+
+In PaddlePaddle, we use graph of operators to describe computation. So the Python class `Optimizer` defines the API to add Ops to computation block. It adds two kinds of Operators:
+
+1. Gradient Ops. Used to calculate the gradients for parameters.
+2. Optimize Ops. Used to apply gradient to parameters.
+
+The python user interface is:
+
+```python
+class Optimizer(object):
+	def _backward(loss):
+		"""
+		Add Operators to Compute gradients of `loss` 
+		It returns the backward net.
+		"""
+		...
+		return backward_net
+
+	def _update(var_list):
+		"""
+		Add Operators to Apply gradients to variables 
+		in var_list. It returns an update `Operator`.
+		"""
+		...
+		return update_op
+
+	def minimize(loss, var_list):
+		"""Add operations to minimize `loss` by updating `var_list`.
+		
+		This method simply combines calls `_backward()` and
+		`_update()`.
+		"""
+		backward_net = _backward(loss)
+		update_op = _update(backward_net)
+		return update_op
+```
+
+because we do not want users to know the step of `_backward` and `_update`, so we decide to export only `minimize()` to users.
```

Comment:
We have backward first, and then update later. And will we implement the stragety that backward and update op by op at the same time?

Reply:
backward will tack a loss as input and add gradient op to all the related variable,  so backward will be a whole step, but optimizer take a parameter `var_list`, so you can add update op to one var at a time and add different update op to different paramets

Reply:
I mean when we run the network, can we have gradients computing and gradients updating in parallel. 
In paddle, we use callback to implement this. Once a Op's backward gradient is calculated, the callback updating function is executed to update the gradient. At the same time, next Op's gradients is calculating.

Reply:
This can be done be re arrange the order of operators in the netop/block when run.